toolchain/check/import_ref.cpp

```diff
@@ -1568,6 +1570,131 @@ class ImportRefResolver {
+  // actually evaluate to themselves to avoid this hack.
+  static auto ImplDeclIdAsConstantId(SemIR::InstId impl_decl_id)
+      -> SemIR::ConstantId {
+    return SemIR::ConstantId::ForTemplateConstant(impl_decl_id);
+  }
+
+  // Gets the instruction ID for an impl declaration represented by the
+  // specified constant ID.
+  auto ConstantIdAsImplDeclId(SemIR::ConstantId impl_const_id)
+      -> SemIR::InstId {
+    return context_.constant_values().GetInstId(impl_const_id);
+  }
+
+  // Make a declaration of an impl. This is done as a separate step from
+  // importing the impl definition in order to resolve cycles.
+  auto MakeImplDeclaration(const SemIR::Impl& import_impl)
+      -> std::pair<SemIR::ImplId, SemIR::ConstantId> {
+    SemIR::ImplDecl impl_decl = {.impl_id = SemIR::ImplId::Invalid,
+                                 .decl_block_id = SemIR::InstBlockId::Empty};
+    auto impl_decl_id =
+        context_.AddPlaceholderInstInNoBlock(context_.MakeImportedLocAndInst(
+            AddImportIRInst(import_impl.latest_decl_id()), impl_decl));
+    impl_decl.impl_id = context_.impls().Add(
+        {GetIncompleteLocalEntityBase(impl_decl_id, import_impl),
+         {.self_id = SemIR::TypeId::Invalid,
+          .constraint_id = SemIR::TypeId::Invalid,
+          .witness_id = SemIR::InstId::Invalid}});
+
+    // Write the impl ID into the ImplDecl.
+    context_.ReplaceInstBeforeConstantUse(impl_decl_id, impl_decl);
+    return {impl_decl.impl_id, ImplDeclIdAsConstantId(impl_decl_id)};
+  }
+
+  // Imports the definition of an impl.
+  auto AddImplDefinition(const SemIR::Impl& import_impl, SemIR::Impl& new_impl,
+                         SemIR::InstId witness_id) -> void {
+    new_impl.definition_id = new_impl.first_owning_decl_id;
+
+    new_impl.witness_id = witness_id;
+
+    // Import the definition scope, if we might need it. Name lookup is never
+    // performed into this scope by a user of the impl, so this is only
+    // necessary in the same library that defined the impl, in order to support
+    // defining members of the impl out of line in the impl file when the impl
+    // is defined in the API file.
+    if (import_ir_id_ == SemIR::ImportIRId::ApiForImpl) {
+      new_impl.scope_id = context_.name_scopes().Add(
+          new_impl.first_owning_decl_id, SemIR::NameId::Invalid,
+          new_impl.parent_scope_id);
+      auto& new_scope = context_.name_scopes().Get(new_impl.scope_id);
+      const auto& import_scope =
+          import_ir_.name_scopes().Get(import_impl.scope_id);
+
+      // Push a block so that we can add scoped instructions to it.
+      context_.inst_block_stack().Push();
+      AddNameScopeImportRefs(import_scope, new_scope);
+      new_impl.body_block_id = context_.inst_block_stack().Pop();
+    }
+  }
+
+  auto TryResolveTypedInst(SemIR::ImplDecl inst,
+                           SemIR::ConstantId impl_const_id) -> ResolveResult {
```

Comment:
Where should I be looking for tests of what's changing here? I'm not seeing more things being tested in the modified `import.carbon` files, and maybe we don't have any tests of `forall` on import boundaries?

What about out-of-line definitions of functions in an impl? (or is that just not supported yet, even though there are some notes around line 1620?)

Reply:
Added check/testdata/impl/no_prelude/import_generic.carbon to test this. In doing so, I found a couple of things that didn't work; fixed. (Though I think the way we're importing generic eval blocks is a bit broken in general and I should have another look at that; I've patched it up enough to get the test passing at least.)

There's currently no support for naming an `impl` as a `NameComponent`, so it's not possible to define a `default fn` of an `impl` out of line yet. (We also don't support `default fn` in general.) So I don't think that can be tested right now. :(